### DISCUSSION TITLE
Vitest watch mode closes instantly after running tests

### DISCUSSION THREAD

**Discussion by nuntax** [vitest-dev/vitest#3890] [Q&A] [answered]
I am currently developing a library to transfer files with the help of webRTC.[https://github.com/nuntax/p2p-files](url)
This is my test code
```js
import { expect, test } from "vitest";
import { FilePeer } from ".";
//@ts-ignore
import wrtc from "wrtc";

let receiver: any;
let sender: any;

test("Connect Peers", async () => {
  receiver = new FilePeer({ wrtc: wrtc, initiator: false, trickle: false });
  sender = new FilePeer({ wrtc: wrtc, initiator: true, trickle: false });
  expect(receiver).toBeDefined();
  expect(sender).toBeDefined();
}, 10000);
```
I am running my tests with `vitest --ui` but after the tests were run vitest just closes and the ui (obviously) cant connect to the vitest server anymore.
The console output looks like the following:

```
> p2p-files@0.0.0 test
> vitest --ui


 DEV  v0.34.1 C:/Users/Name/Desktop/Project/p2plibrary/p2p-files
      UI started at http://localhost:51204/__vitest__/

 ✓ src/lib.test.ts (1)
   ✓ Connect Peers
```

**Comment by nuntax** [comment]
What I've discovered after is that watch mode starts up normally without the initialization of sender and receiver. I can also add them and the test runs without Problem. However as soon as I change something and the test reruns I am getting an Error message that I can't see because vitest is exiting before.
I have 2 questions now.
1: Should I write a singleton file which exports both sender and receiver and checks if they already exist so I can destroy them in that case. So they can be reinitiated
2: Is there a way to make vitest output a log file which I could use to debug the issue and report the underlaying bug or fix it and make a PR?

**Reply by sheremet-va** [reply, marked as answer]
1. You should probably cleanup after your test (remove all listeners) using `afterEach`
2. There is no log unfortunately

**Reply by scott-cornwell** [reply]
@nuntax I have a very similar problem. I created a new nx workspace and had it set up vitest for me. I was able to add a new target for watch mode but runs the test, then watches for file changes and exits immediately with no errors. I don't even have any real tests, just the one included dummy test.

**Reply by hi-ogawa** [reply]
@scott-cornwell `nx` uses its own wrapper for Vitest execution. Unless you can reproduce the issue with the official `vitest` cli, it's better to be reported on nx. For example, you might already find some known issues in https://github.com/nrwl/nx/issues?q=is%3Aissue+is%3Aopen+vitest

**Comment by Kasopej** [comment]
Hi @sheremet-va 
your solution is straight forward for cases where dev is not tearing down

But the issue with vitest closing watch mode still persists for me. I do not seem to have issues with tearing down/state leaks, just the vitest process ending
I initially had issues with msw and node 18 native fetch bug in thread processing mode (see https://github.com/nodejs/undici/issues/2026), but switching to forks mode has solved that for me

But I would love the vitest watch process to stay alive as that is the whole point of me keeping tests running in development

**Comment by cinderblock** [comment]
I came across this issue because I was having the same symptoms.

Turns out I was running `vitest run --ui` instead of `vitest --ui`.

Hope this helps someone else.
